lib/hmmlearn/hmm.py

```diff
@@ -307,18 +309,246 @@ def _do_mstep(self, stats):
+            combination of 's' for startprob, 't' for transmat, and 'e' for
+            emissionprob.  Defaults to all parameters.
+        
+        implementation: string, optional
+            Determines if the forward-backward algorithm is implemented with
+            logarithms ("log"), or using scaling ("scaling").  The default is
+            to use logarithms for backwards compatability.
+        """
+        BaseHMM.__init__(self, n_components,
+                          startprob_prior=startprob_prior,
+                          transmat_prior=transmat_prior,
+                          algorithm=algorithm,
+                          random_state=random_state,
+                          n_iter=n_iter, tol=tol, verbose=verbose,
+                          params=params, init_params=init_params,
+                          implementation=implementation)
+        self.n_trials = n_trials
+        self.n_features = n_features
+        self.emission_prior = emission_prior
+        
+        _log.warning("MultinomialHMM has undergone major changes. "
+                "The previous version was actually implenting CategoricalHMM. This new "
+                "implementation follows the standard definition for a Multinomial distribution, "
+                "e.g. as in https://en.wikipedia.org/wiki/Multinomial_distribution"
+                "See these issues for details: "
+                "\nhttps://github.com/hmmlearn/hmmlearn/issues/335 "
+                "\nhttps://github.com/hmmlearn/hmmlearn/issues/340 "
+                "\nThis is also why n_components and n_trials are now required arguments.")
+
+    # From the API: Return a mapping of fittable parameter names 
+    # (as in self.params) to the number of corresponding scalar parameters 
+    # that will actually be fitted. This is used to detect whether the user 
+    # did not pass enough data points for a non-degenerate fit.
+    def _get_n_fit_scalars_per_param(self):
+        nc = self.n_components
+        nf = self.n_features
+        return {
+            "s": nc - 1,
+            "t": nc * (nc - 1),
+            "e": nc * (nf - 1),
+        }
+
+    
+    def _check_and_set_multinomial_n_features(self, X):
+        """
+        Check if ``X`` is a sample from a multinomial distribution, i.e. an
+        array of non-negative integers, summing up to n_trials.
+        """
+        n_samples, n_features = X.shape
+        if n_features != self.n_features:
+            raise ValueError("Feature dimension should be (n_samples, n_features)")
+        if not (X.sum(axis=1) == self.n_trials).all():
+            raise ValueError("Total count for each sample should add up to number of trials")
+        if not np.issubdtype(X.dtype, np.integer):
+            raise ValueError("Symbols should be integers")
+        if X.min() < 0:
+            raise ValueError("Symbols should be nonnegative")
+        if hasattr(self, "n_features"):
+            if self.n_trials < X.max():
```

Comment:
I think this condition redundant as it is already checked by `X.sum(-1) == n_trials and all(X >= 0)` ?

Reply:
Makes sense, edited.